src/Nest/Document/Multiple/Reindex/ElasticClient-Reindex.cs

```diff
@@ -21,59 +57,126 @@ public partial interface IElasticClient
+		/// <param name="selector">an optional query limitting the documents found in <paramref name="fromIndex"/></param>
+		IObservable<IBulkAllResponse> Reindex<TSource>(
+			IndexName fromIndex,
+			IndexName toIndex,
+			Func<QueryContainerDescriptor<TSource>, QueryContainer> selector = null,
 			CancellationToken cancellationToken = default(CancellationToken)
-		) where T : class;
+		)
+			where TSource : class;
+
 	}
 
 	public partial class ElasticClient
 	{
 		/// <inheritdoc />
-		public IObservable<IBulkAllResponse> Reindex<T>(
-			Func<ReindexDescriptor<T>, IReindexRequest<T>> selector,
+		public IObservable<IBulkAllResponse> Reindex<TSource>(
+			Func<ReindexDescriptor<TSource,TSource>, IReindexRequest<TSource,TSource>> selector,
 			CancellationToken cancellationToken = default(CancellationToken)
-		) where T : class =>
-			this.Reindex<T>(selector.InvokeOrDefault(new ReindexDescriptor<T>()));
+		)
+			where TSource : class =>
+			this.Reindex<TSource,TSource>(selector.InvokeOrDefault(new ReindexDescriptor<TSource,TSource>(s=>s)));
 
 		/// <inheritdoc />
-		public IObservable<IBulkAllResponse> Reindex<T>(
-			IndexName fromIndex,
-			IndexName toIndex,
-			Func<QueryContainerDescriptor<T>, QueryContainer> selector = null,
+		public IObservable<IBulkAllResponse> Reindex<TSource,TTarget>(
+			Func<TSource, TTarget> mapper,
+			Func<ReindexDescriptor<TSource,TTarget>, IReindexRequest<TSource,TTarget>> selector,
 			CancellationToken cancellationToken = default(CancellationToken)
-		) where T : class =>
-			this.Reindex<T>(r => r
-					.ScrollAll("1m", -1, search => search.Search(ss => ss.Index(fromIndex).Query(selector)))
-					.BulkAll(b => b.Index(toIndex))
-			, cancellationToken);
+		)
+			where TTarget : class
+			where TSource : class =>
+			this.Reindex<TSource,TTarget>(selector.InvokeOrDefault(new ReindexDescriptor<TSource,TTarget>(mapper)));
+
+		/// <inheritdoc />
+		public IObservable<IBulkAllResponse> Reindex<TSource>(
+			IReindexRequest<TSource> request,
+			CancellationToken cancellationToken = default(CancellationToken)
+		)
+			where TSource : class =>
+			this.Reindex<TSource, TSource>(request, cancellationToken);
 
 
 		/// <inheritdoc />
-		public IObservable<IBulkAllResponse> Reindex<T>(
-			IReindexRequest<T> request,
+		public IObservable<IBulkAllResponse> Reindex<TSource,TTarget>(
+			IReindexRequest<TSource,TTarget> request,
 			CancellationToken cancellationToken = default(CancellationToken)
-		) where T : class
+		)
+			where TTarget : class
+			where TSource : class
 		{
 			if (request.ScrollAll == null)
 				throw new ArgumentException(
 					"Reindex must have ScrollAll property set so we know how to get the source of the reindex operation");
 			if (request.BulkAll == null)
 				throw new ArgumentException(
 					"Reindex must have BulkAll property set so we know how to get the target of the reindex operation");
```

Comment:
minor grammar nitpick: I think we should remove the subject pronouns in exception messages, for example
instead of:

> Reindex must have BulkAll property set so we know how to get the target of the reindex operation

can be

> BulkAll property must set in order to get the target of a Reindex operation